### ISSUE TITLE
Tool installation failure due to dependency history

### ISSUE THREAD

**Issue by yhoogstrate** [galaxyproject/galaxy#667]
I think I am experiencing a bug. A Galaxy server remembers and expects dependencies of older revisions.

Example use case: In a toolshed I have a repository of a tool, say package_new_gene_db_1_2_3, and in revision 1 it depends on package_sqlite_1_0_0:
- package_new_gene_db_1_2_3
  - package_sqlite_1_0_0

At some point in time I discover that postgres is a better solution for my gene database, and I upload revision 2:
- package_new_gene_db_1_2_3
  - package_postgres__2_0_0

And I remove the dependecy of sqlite.

If I go to my galaxy instance and I remove revision 1 and install revision 2, it (often/always?) happens that the installation of revision 2 fails, because package_sqlite_1_0_0 can not be found, while it doesn't depend on it anymore. After double checking the entries in the tool sheds, I am absolutely certain that there is no dependency on sqlite anymore in the tool_dependencies.xml.

When I use a clean instance of galaxy on which revision 1 was not installed, and I immediately install revision 2, this bug does never appear. Because the problem only occurs on servers on which older revisions of tools were installed, I expect that this is a bug in galaxy and not in the toolshed.

What did not work is uninstalling and reinstalling all the dependencies, the problem keeps coming back.

An example is package_python_2_7. It once depended on package_openssl_1_0 which has now been deprecated, but all tools that depend on package_python_2_7 end up gray. Still, I get errors that package_openssl_1_0 is not installed. Installing package_openssl_1_0 doesn't help either, because the installation fails, which is why it was deprecated.

I experienced this bug also with other tools, on an old and new (pulled yesterday) instance of galaxy.

Could it be the same bug as issue https://github.com/galaxyproject/galaxy/issues/525 ?
**[edit] The issue is still present after that one was patched, so it's a different problem**

![screenshot from 2015-08-28 14 24 30](https://cloud.githubusercontent.com/assets/2264697/9546523/9f14a002-4d93-11e5-8279-2a131aaae0ff.png)

![screenshot from 2015-08-28 14 26 28](https://cloud.githubusercontent.com/assets/2264697/9546526/a5cc3900-4d93-11e5-9ed9-201542bc2a53.png)

**Comment by yhoogstrate** [comment]
I found another issue related to comment tags in tool_dependencies.xml.
Comments inside the *<package />* tag break the installation of tools, resulting in the following error:

*Error installing tool dependency <tool name>: 'NoneType' object has no attribute process_tag_set*

An example is given in the following 2 figures:

![xml](https://cloud.githubusercontent.com/assets/2264697/9578487/b3f9b4d6-4fea-11e5-9d52-f76a23c3ddbe.png)
![error](https://cloud.githubusercontent.com/assets/2264697/9578490/b8181710-4fea-11e5-88cf-6afd60adc7ff.png)

**Comment by mvdbeek** [comment]
I also ran into this problem (with PR #681 applied).
Considering I know the code now a bit I'll dig into it.

**Comment by yhoogstrate** [comment]
Awesome! I hope you are able to fix, or give some clues.

**Comment by mvdbeek** [comment]
So i first worked on making this reproducible. The archive below contains a number of toy packages. https://bitbucket.org/mvdbeek/galaxy_downloads/downloads/fix_unrelated_repo_install.tgz
If you want to reproduce the bug, you need to upload the packages to a toolshed, in the the order package_fixed, package_broken, package_switch1 and then install package_switch.
package_switch now depends on package_fixed. All packages should be properly installed. Uninstall all packages from the galaxy instance. Next, upload package_switch2 to the same toolshed (which is the same package as package_switch1, but now depends on package_broken). Then install package_switch, and you will see that the repository install fails because package_fixed is not installed (which it shouldn't be).

The problem is that during the install galaxy first tries to get the repositories to be installed from the database. This works fine if the changeset revisions are the same, or you add new dependencies in a new changeset revision, but it doesn't work if dependencies are removed. I "fixed" this (PR #707) by making sure that the dependencies retrieved from the database are checked against the dependencies of the repository that is to be installed. If a dependency is present in the database but not in self, it will not be added to the list of misssing repository and tool dependencies

I do think however that this needs to be carefully checked and reworked by someone who knows more about the toolshed, because I doubt that I made the check in the correct place and that the comparison to self always works.

I tested this with the above approach, and this is working correctly now, as well as with package_superswitch, which depends on package_switch, which depends on either package_broken or package_fixed. I hope this will help someone else to fix it properly.

**Comment by mvdbeek** [comment]
Adding to this bug report, similar behavior occurs if the owner or changeset_revision of a dependent package changes.
So in the above example uninstalling package switch and package_fixed (for simplicity: rev. 0000001), then updating package_fixed (with or without new owner), then uploading new versions of package_fixed and package_switch and installing package_switch leads to package_fixed rev. 0000002 being correctly installed, however package_switch will have a missing repository dependency of package_fixed rev. 0000001.
I addressed this in my last commit 6b96717 to the PR.

@yhoogstrate, by any chance, could you confirm that PR #707 fixes the behavior you described?
It should be sufficient to just apply the patch and restart galaxy, no re-install of tools/packages required.

**Comment by mvdbeek** [comment]
@erasche I think this merits a bug label.

**Comment by hexylena** [comment]
@mvdbeek sorry, had put a pause on my triaging/tagging activities as I needed clarifications on tag meanings. Edit: yep that sounds like a bug.

**Comment by mvdbeek** [comment]
haha, yeah, no problemo.

**Comment by mvdbeek** [comment]
Another report from the mailing list here:
http://dev.list.galaxyproject.org/recurring-issues-with-revisions-of-packages-from-toolshed-td4668101.html

**Comment by yhoogstrate** [comment]
This one is really bugging one of our production servers because package_python_2_7 once depended on openssl. The only way to get around is to take the server down and reinstall, which we really don't want to do.

I am not so much into this part of the code so maybe you can shed some light on this @mvdbeek? Is there no way this can be fixed when all packages get uninstalled and reinstalled? At the moment, if all packages are being removed, the bug keeps coming back when the tools get re-instsalled, so something is not entirely removed. It wouldn't solve the complete issue but it would at least be a workaround.

@jmchilton I saw you wrote a piece of the code that @mvdbeek suspects to be involved in it. Maybe you have a clue?

**Comment by mvdbeek** [comment]
Well, you can uninstall package_python_2_7, then that database entry is being set to deleted.
Then you'd need to remove that entry from the database, and then you can reinstall without fetching the old dependency. I haven't tested this script, but it looks good:
https://gist.github.com/Unode/9e64bb1f9b7ad930d44f

**Comment by bgruening** [comment]
@yhoogstrate can you please test https://github.com/galaxyproject/galaxy/pull/1154? Thanks!

**Comment by yhoogstrate** [comment]
Absolutely, will keep you posted!

**Comment by yhoogstrate** [comment]
For this test I used git revision:

```
commit 11f9fed4acb3fb912bc0e0e5c0232fb85c974f2f
Merge: 2e510bd a16cb06
Author: Martin Cech <cech.marten@gmail.com>
Date:   Sun Nov 22 20:12:00 2015 -0500

    Merge pull request #1154 from gregvonkuster/dev

    Fix for re-installing an uninstalled TS repository with a dependency
```

With @mvdbeek's test-data, after switching from fixed to broken and back, I get the following behavior (please neglect that _fixed_ is failing because I broke it):

![screenshot from 2015-11-24 16-04-24](https://cloud.githubusercontent.com/assets/2264697/11370832/b8e764a2-92c5-11e5-8874-0d8f7e2fe840.png)

_fixed_ is not a missing tool dependency anymore, which is a good thing. I am not sure if the reference to package_fixed is okay.

---

If I then make _package_fixed_ work again and switch _package_switch_ to use package_fixed, I get the following:
![screenshot from 2015-11-24 16-22-31](https://cloud.githubusercontent.com/assets/2264697/11371348/57b22750-92c8-11e5-87c5-89e2cf0de5b8.png)
While both dependencies are correctly installed the tool is marked as missing dependencies. I don't think PR #1154 entirely solves the issue.

**Comment by gregvonkuster** [comment]
Thanks for testing this!  The scenario you describe above, specifically the last section, is definitely different than the scenario I used to implement the fix in my PR.  The fix for the above issue will have to be in a separate PR.

**Comment by mvdbeek** [comment]
@yhoogstrate this is solved for me with PR #1185

**Comment by martenson** [comment]
fixed  by #1193 
big thanks to @mvdbeek for detailed report and @gregvonkuster for fixing it!
